src/DataProvider.js

```diff
@@ -72,6 +72,23 @@ export default class DataProvider {
     }, this.polling())
   }
 
+  getDataWithRetry() {
+    return new Promise((resolve, reject) => {
+      let lastTimeout
+      let timedGetData = async (retries) => {
+        if (retries <= 0) {
+          reject()
+          return
+        }
+        lastTimeout = setTimeout(timedGetData, cfg.fetchTimeout, retries - 1)
+        let data = await this.getData()
+        clearTimeout(lastTimeout)
```

Comment:
Hmm, why `setTimeout` and not our nice cancellable bluebird Promises?

Also, what happens if you call `resolve` more times, on a resolved Promise? I think it'd be best to ensure that it's resolved only once.

Perhaps you could make a list of delayed promises and make them race? Just a suggestion.

Reply:
Concerning the 2nd point - calling `resolve` multiple times - only the first one is important, others are ignored. (I've tested tested by using different delays in `getData` implementation, e.g. 5000 ms for 1st req, 3000 for 2nd, and 1000 for 3rd req).

I thought about the list of promises and `Promise.race` solution, but it seemed like an overkill to create `maxTimeoutRetries` promises for every request, even though they may not be needed (no timeouts happening). This solution seems more optimal in that regard, since it creates a new Promise only when the previous times out.